spacy/scorer.py

```diff
@@ -473,17 +473,17 @@ def score_cats(
             threshold = 0.5 if multi_label else 0.0
         if not multi_label:
             threshold = 0.0
+        labels = set(labels)
         f_per_type = {label: PRFScore() for label in labels}
         auc_per_type = {label: ROCAUCScore() for label in labels}
-        labels = set(labels)
-        if labels:
-            for eg in examples:
-                labels.update(eg.predicted.cats.keys())
-                labels.update(eg.reference.cats.keys())
         for example in examples:
             # Through this loop, None in the gold_cats indicates missing label.
-            pred_cats = getter(example.predicted, attr)
-            gold_cats = getter(example.reference, attr)
+            pred_cats = {
+                k: v for k, v in getter(example.predicted, attr).items() if k in labels
+            }
+            gold_cats = {
+                k: v for k, v in getter(example.reference, attr).items() if k in labels
+            }
```

Comment:
As these are all the same edits as in https://github.com/explosion/spaCy/pull/11820/files, I would have preferred to continue working from that PR rather than open a parallel one. We can remove the merging of the dict and keep that in a separate PR / for a separate discussion / more thorough solution, but this part is pretty much the same.

Reply:
Sure, I can move these changes over there.